### PR TITLE
Big UI Stuff & Other stuff

### PR DESCRIPTION
**Pull request by TheGrimsey**
**Changes proposed:** (Add / Remove lines if necessary)
- Depth Sorting
- Widget Visibility
- SceneManager in Angelscript
- DataStorage in Angelscript
- Make use of empty components for stuff.
- Allow deleting widgets
- Replenish pool when it runs out.
- Clean up includes at random places

**Tests Performed:** (Add / Remove lines if necessary)

- [X] Build
- [X] Tested in-game

**Known issues & TODO list** (Add / Remove lines if necessary)
- Depth sorting between parents and children isn't quite there 
- Bounds are not finished.

**Targeted branch(es):**
- [x] Master

### CODE REVIEW

File: client/ECS/Components/UI/UIVisiblity.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+struct UIVisiblity
```
Comment (Pursche) [inline client/ECS/Components/UI/UIVisiblity.h:3]: Visiblity -> Visibility please, and in the file name as well.

File: client/Scripting/Classes/UI/asUITransform.cpp [hunk 2]
```diff
@@ -4,6 +4,8 @@
 #include "../../../ECS/Components/UI/UIDataSingleton.h"
 #include "../../../ECS/Components/UI/UITransformUtils.h"
 
+#include "../../../ECS/Components/UI/UIVisible.h"
```
Comment (Pursche) [inline client/Scripting/Classes/UI/asUITransform.cpp:8]: Unnecessary line break between UIVisible and UITransformUtils?

File: client/Scripting/Classes/UI/asUITransform.cpp [hunk 3]
```diff
@@ -219,6 +244,39 @@ namespace UI
...
+
+        // TRANSACTION
+        entt::registry* gameRegistry = ServiceLocator::GetGameRegistry();
+        entt::entity entId = _entityId;
+        gameRegistry->ctx<ScriptSingleton>().AddTransaction([visible, entId]()
+            {
+                entt::registry* uiRegistry = ServiceLocator::GetUIRegistry();
+                UIVisiblity& uiVisibility = uiRegistry->get<UIVisiblity>(entId);
+
+                const bool visiblityChanged = uiVisibility.visible != visible;
```
Comment (Pursche) [inline client/Scripting/Classes/UI/asUITransform.cpp:318]: Double check visiblity -> visibility for all files please

File: client/ECS/Components/UI/Singletons/UIDataSingleton.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        entt::registry* registry = ServiceLocator::GetUIRegistry();
+
+        std::vector<entt::entity> entityIds = std::vector<entt::entity>();
+        entityIds.reserve(entityToAsObject.size());
+
+        for (auto asObject : entityToAsObject)
+        {
+            entityIds.push_back(asObject.first);
+
```
Comment (NixAJ) [inline client/ECS/Components/UI/Singletons/UIDataSingleton.cpp:19]: No need to have a space here

File: client/ECS/Components/UI/Singletons/UIEntityPoolSingleton.cpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#include "UIEntityPoolSingleton.h"
+#include "../../../../Utils/ServiceLocator.h"
+
+namespace UI
+{
+    entt::entity UIEntityPoolSingleton::DeQeueueId()
```
Comment (NixAJ) [inline client/ECS/Components/UI/Singletons/UIEntityPoolSingleton.cpp:6]: Call this GetId

File: client/ECS/Components/UI/Singletons/UIEntityPoolSingleton.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        entt::entity entityId;
+
+        if (!entityIdPool.try_dequeue(entityId))
+        {
+            AllocatePool();
+
```
Comment (NixAJ) [inline client/ECS/Components/UI/Singletons/UIEntityPoolSingleton.cpp:13]: Unneccesary new line

File: client/ECS/Components/UI/Singletons/UIEntityPoolSingleton.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    struct UIEntityPoolSingleton
+    {
+        const int ENTITIES_TO_PREALLOCATE = 10000;
+
+    public:
+        UIEntityPoolSingleton() : entityIdPool(ENTITIES_TO_PREALLOCATE) { }
+
+        entt::entity DeQeueueId();
+
+        void AllocatePool();
```
Comment (NixAJ) [inline client/ECS/Components/UI/Singletons/UIEntityPoolSingleton.h:16]: Place this above GetId and no need to newline

File: client/ECS/Systems/UI/AddElementSystem.cpp [hunk 8]
```diff
@@ -1,35 +1,49 @@
...
+#include "../../Components/UI/Singletons/UIAddElementQueueSingleton.h"
 #include "../../Components/UI/UITransform.h"
 #include "../../Components/UI/UITransformEvents.h"
+
 #include "../../Components/UI/UIRenderable.h"
+#include "../../Components/UI/UIImage.h"
 #include "../../Components/UI/UIText.h"
+
+#include "../../Components/UI/UICollision.h"
+
```
Comment (NixAJ) [inline client/ECS/Systems/UI/AddElementSystem.cpp:15]: No need for new line here

File: client/Rendering/UIRenderer.cpp [hunk 9]
```diff
@@ -1,140 +1,127 @@
...
-                else // Otherwise we just update the already existing primitive model
-                {
-                    _renderer->UpdatePrimitiveModel(renderable.modelID, primitiveModelDesc);
-                }
-
-                transform.isDirty = false;
+                _renderer->UpdatePrimitiveModel(image.modelID, primitiveModelDesc);
             }
         });
+    registry->remove<UIDirty>(renderableView.begin(), renderableView.end());
```
Comment (NixAJ) [inline client/Rendering/UIRenderer.cpp:117]: Considering we clear all UIDirty components in this function, perhaps move the remove with iterators, to just destroy all of them at the very end?

File: client/Rendering/UIRenderer.cpp [hunk 10]
```diff
@@ -221,10 +208,8 @@ void UIRenderer::Update(f32 deltaTime)
             text.constantBuffer->resource.outlineWidth = text.outlineWidth;
             text.constantBuffer->Apply(0);
             text.constantBuffer->Apply(1);
-
-            transform.isDirty = false;
-            text.isDirty = false;
         });
+    registry->remove<UIDirty>(textView.begin(), textView.end());
```
Comment (NixAJ) [inline client/Rendering/UIRenderer.cpp:212]: Considering we clear all UIDirty components in this function, perhaps move the remove with iterators, to just destroy all of them at the very end?

File: client/Scripting/Classes/SceneManager/SceneManagerUtils.cpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
+#include "SceneManagerUtils.h"
+#include "../../ScriptEngine.h"
+
+#include "../../../Utils/ServiceLocator.h"
+#include "../scenemanager-lib/SceneManager.h"
```
Comment (NixAJ) [inline client/Scripting/Classes/SceneManager/SceneManagerUtils.cpp:5]: This is bad

File: client/Scripting/Classes/UI/asInputfield.cpp [hunk 12]
```diff
@@ -1,11 +1,13 @@
...
 #include "../../../Utils/ServiceLocator.h"
-#include "../../../ECS/Components/UI/UIEntityPoolSingleton.h"
+
+#include "../../../ECS/Components/UI/Singletons/UIEntityPoolSingleton.h"
 #include "../../../ECS/Components/Singletons/ScriptSingleton.h"
-#include "../../../ECS/Components/UI/UIAddElementQueueSingleton.h"
+
+#include "../../../ECS/Components/UI/UIDirty.h"
+
+#include <GLFW/glfw3.h>
```
Comment (NixAJ) [inline client/Scripting/Classes/UI/asInputfield.cpp:10]: Is this really needed?

File: client/Scripting/Classes/UI/asInputfield.cpp [hunk 13]
```diff
@@ -42,21 +44,50 @@ namespace UI
...
+        case GLFW_KEY_LEFT:
+            MovePointerLeft();
+            break;
+        case GLFW_KEY_RIGHT:
+            MovePointerRight();
+            break;
+        case GLFW_KEY_ENTER:
+        {
+            entt::registry* registry = ServiceLocator::GetUIRegistry();
+            registry->get<UIInputField>(GetEntityId()).OnSubmit();
```
Comment (NixAJ) [inline client/Scripting/Classes/UI/asInputfield.cpp:66]: Perhaps put GetEntityId into a temporary variable

File: client/Scripting/Classes/UI/asPanel.cpp [hunk 14]
```diff
@@ -86,48 +88,41 @@ namespace UI
...
-        _renderable.texture = texture;
+        _image.texture = texture;
 
         entt::registry* gameRegistry = ServiceLocator::GetGameRegistry();
         entt::entity entId = _entityId;
         gameRegistry->ctx<ScriptSingleton>().AddTransaction([texture, entId]()
             {
                 entt::registry* uiRegistry = ServiceLocator::GetUIRegistry();
-                UIRenderable& renderable = uiRegistry->get<UIRenderable>(entId);
+                UIImage& renderable = uiRegistry->get<UIImage>(entId);
```
Comment (NixAJ) [inline client/Scripting/Classes/UI/asPanel.cpp:98]: Update name to reflect what it is

File: client/Scripting/Classes/UI/asPanel.cpp [hunk 15]
```diff
@@ -86,48 +88,41 @@ namespace UI
...
-        _renderable.color = color;
+        _image.color = color;
 
         entt::registry* gameRegistry = ServiceLocator::GetGameRegistry();
         entt::entity entId = _entityId;
         gameRegistry->ctx<ScriptSingleton>().AddTransaction([color, entId]()
             {
                 entt::registry* uiRegistry = ServiceLocator::GetUIRegistry();
-                UIRenderable& renderable = uiRegistry->get<UIRenderable>(entId);
+                UIImage& renderable = uiRegistry->get<UIImage>(entId);
```
Comment (NixAJ) [inline client/Scripting/Classes/UI/asPanel.cpp:113]: Update name to reflect what it is

File: client/Scripting/Classes/UI/asUITransform.cpp [hunk 16]
```diff
@@ -219,36 +297,168 @@ namespace UI
...
+    }
+
+    void asUITransform::SetCollisionEnabled(bool enabled)
+    {
+        entt::registry* gameRegistry = ServiceLocator::GetGameRegistry();
+        entt::entity entId = _entityId;
+        gameRegistry->ctx<ScriptSingleton>().AddTransaction([enabled, entId]()
+            {
+                entt::registry* uiRegistry = ServiceLocator::GetUIRegistry();
+                if (uiRegistry->has<UICollision>(entId) == enabled)
```
Comment (NixAJ) [inline client/Scripting/Classes/UI/asUITransform.cpp:340]: Possibly add a comment here since it was a bit confusing to read

**Pull request merged**